linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala

```diff
@@ -0,0 +1,1496 @@
+              List(zero, str(className), str(displayName), envField("isArrayOf", className)))
+        })
+      }
+    }
+
+    private def maybeWrapInUBE(behavior: CheckedBehavior, exception: Tree): Tree = {
+      if (behavior == CheckedBehavior.Fatal)
+        genScalaClassNew("sjsr_UndefinedBehaviorError", "init___jl_Throwable", exception)
+      else
+        exception
+    }
+
+    private def genScalaClassNew(className: String, ctorName: String,
+        args: Tree*): Tree = {
+      Apply(New(encodeClassVar(className), Nil) DOT ctorName, args.toList)
+    }
+
+    private def genIsScalaJSObject(obj: VarRef): Tree =
+      !(!(obj && (obj DOT classData)))
+
+    private def genIsScalaJSObjectOrNull(obj: VarRef): Tree =
+      genIsScalaJSObject(obj) || (obj === Null())
+
+    private def envFunctionDef(name: String, args: List[ParamDef],
+        body: Tree): FunctionDef = {
+      FunctionDef(envFieldIdent(name), args, body)
+    }
+
+    private def genClassDef(className: Ident, parent: Option[(Tree, Tree)],
+        members: List[Tree]): Tree = {
+      if (useECMAScript2015) {
+        ClassDef(Some(className), parent.map(_._1), members)
+      } else {
+        val classRef = VarRef(className)
+
+        val ctor = members.collectFirst {
+          case m @ MethodDef(false, Ident("constructor", _), _, _) => m
+        }.getOrElse {
+          MethodDef(false, Ident("constructor"), Nil, Skip())
+        }
+
+        val patchedCtorBody = {
+          def patchSuper(stat: Tree): Tree = stat match {
+            case Apply(Super(), args) =>
+              /* All the super constructor calls we have in the core JS lib are
+               * actually no-op, so we just omit them.
+               */
+              assert(args.isEmpty)
+              Skip()
+            case _ =>
+              stat
+          }
+
+          ctor.body match {
+            case Block(stats) => Block(stats.map(patchSuper(_)))
+            case stat         => patchSuper(stat)
+          }
+        }
+
+        val ctorFun = FunctionDef(className, ctor.args, patchedCtorBody)
+
+        val prototype = classRef.prototype
+
+        val inheritProto = parent.fold[Tree] {
+          Skip()
+        } { parent =>
+          val inheritableCtor = parent._2
+          Block(
+              prototype := New(inheritableCtor, Nil),
+              (prototype DOT "constructor") := classRef
+          )
+        }
+
+        val setMembers = for (member <- members) yield {
+          (member: @unchecked) match {
+            case MethodDef(false, Ident("constructor", _), _, _) =>
+              Skip()
+
+            case MethodDef(static, name, args, body) =>
+              val target = if (static) classRef else prototype
+              genPropSelect(target, name) := Function(arrow = false, args, body)
+          }
+        }
+
+        Block(ctorFun :: inheritProto :: setMembers)
+      }
+    }
+
+    private def varRef(name: String): VarRef = VarRef(Ident(name))
```

Comment:
Have you considered using symbols with implicit conversions for this? It might quite shorten the file (and maybe even aid readability).